### PR TITLE
bsdlib: make sure bsdlib dependent services handle shutdown and init

### PR DESCRIPTION
**Pull request by evenl**
I am working on an update to bsdlib to make it possible to run shutdown and init in multiple subsequent sequences. This has some implications for some of the services in NCS that depends on bsdlib because a shutdown of bsdlib will invalidate any socket used by and component in NCS. A shutdown will wake all threads that is sleeping on a socket (in a recv call for instance) and return EHOSTDOWN. 

* The thread in the at_cmd service will go to sleep on bsdlib_shutdown and wakeup and resume operation in bsdlib_ini.
* The thread in the gps driver will go to sleep on bsdlib_shutdown and wakeup on bsdlib_init(). It the GPS driver was in started started it will be reset to initialized state. 
* The thread in the download client will be terminated on bsdlib_shutdown.

Signed-off-by: Even Falch-Larsen <even.falch-larsen@nordicsemi.no>

### CODE REVIEW

File: lib/at_cmd/at_cmd.c [hunk 1]
```diff
@@ -160,7 +162,10 @@ static void socket_thread_fn(void *arg1, void *arg2, void *arg3)
 
 			LOG_ERR("Unrecoverable reception error (err: %d), "
 				"thread killed", errno);
+
+			atomic_clear_bit(&running, 0);
 			close(common_socket_fd);
+			k_mem_slab_free(&rsp_work_items, (void **)&item);
 			return;
```
Comment (rlubos) [inline lib/at_cmd/at_cmd.c:169]: I need to understand, this can only happen in case this thread failed to recover the socket on its own (by simply closing and reopening it). Does it really help if we do the same manually with a call to `at_cmd_driver_init`? How can the application tell that this thread is down and it needs to re-initialize it with `at_cmd_driver_init`?

Comment (evenl) [reply lib/at_cmd/at_cmd.c:169]: The recover attempt will fail because during the shutdown process the interfaces (the different socket types) in bsdlib is brought which cocecently leads to the open() call to fail. So to avoid having a mem slab in allocated state I need to free it here so the whole pool is in free state. 
Actually I think the close on line 167 could be removed, it will anyway be forced shut by the shutdown process.

Comment (lemrey) [reply lib/at_cmd/at_cmd.c:169]: You cannot restart a thread :)
You'll need to run at beginning of the thread function when you expect to be able to open a socket again.

Comment (rlubos) [reply lib/at_cmd/at_cmd.c:169]: Just wondering how should we handle this problem. Assuming that some operation will fail when we disable bsdlib, so we can recover later seems error-prone to me.

I think we should provide `*_deinit` function for all modules that depend on bsdlib to operate (`at_cmd` and all modules that rely on it). This way, an application could do a graceful shutdown of all of the dependent modules, before shutting down bsdlib.

And as @lemrey says, we should not try to respawn a thread. While this might work, it's not a functionality supported by the kernel, and thus can lead to some undefined behaviour. I've finally found some docs to back me up here:
https://docs.zephyrproject.org/latest/reference/kernel/threads/index.html#thread-termination
i.e. 
> The kernel does not currently make any claims regarding an application’s ability to respawn a thread that terminates.

Comment (evenl) [reply lib/at_cmd/at_cmd.c:169]: > Just wondering how should we handle this problem. Assuming that some operation will fail 
> when we disable bsdlib, so we can recover later seems error-prone to me.

I inject that error into the socket on shutdown to signal that the link is down and then I wake all sleeping threads to make them process this, so it's not an assumption. 

But ok, I see, I was assuming that the kernel was freeing the threads context and that it could be respawned after it had died. Seems like I can use the k_thread_join() function and join the thread with a shutdown thread in the os integration code that can exit on the next init. This way I don't need to exit this thread and can make it continue on my next init. The pid to that thread could be passed through a deinit function.

I can make a try and update the PR later today.

File: lib/at_cmd/at_cmd.c [hunk 2]
```diff
@@ -308,10 +323,9 @@ void at_cmd_set_notification_handler(at_cmd_handler_t handler)
 
 static int at_cmd_driver_init(struct device *dev)
 {
-	static bool initialized;
-
-	if (initialized) {
-		return 0;
+	if (atomic_test_bit(&running, 0)) {
```
Comment (rlubos) [inline lib/at_cmd/at_cmd.c:326]: Do we need this change? `at_cmd` does not need to be re-initialized anymore. Please keep the diff minium, avoid unnecessary changes.

File: lib/at_cmd/at_cmd.c [hunk 3]
```diff
@@ -145,6 +147,7 @@ static void socket_thread_fn(void *arg1, void *arg2, void *arg3)
 
 		bytes_read = recv(common_socket_fd, item->data,
 				  sizeof(item->data), 0);
+
```
Comment (rlubos) [inline lib/at_cmd/at_cmd.c:150]: Empty line not needed.

File: include/modem/bsdlib.h [hunk 4]
```diff
@@ -22,6 +22,8 @@ extern "C" {
  */
 int bsdlib_init(void);
 
+void bsdlib_shutdown_wait(void);
```
[the remark this answers is not in the document]
Comment (evenl) [reply include/modem/bsdlib.h:25]: yes, forgot about that, just wanted to push out a new proposal to get get som progress. I'll add documentation.

File: lib/bsdlib/bsdlib.c [hunk 5]
```diff
@@ -39,9 +53,42 @@ static int _bsdlib_init(struct device *unused)
...
 	return init_ret;
 }
 
+void bsdlib_shutdown_wait(void)
+{
+	struct shutdown_thread thread;
+
+	k_sem_init(&thread.sem, 0, 1);
+
+	u32_t key = irq_lock();
```
[the remark this answers is not in the document]
Comment (evenl) [reply lib/bsdlib/bsdlib.c:77]: ok

File: lib/bsdlib/bsdlib.c [hunk 6]
```diff
@@ -39,9 +53,42 @@ static int _bsdlib_init(struct device *unused)
...
+	if (sys_slist_peek_head(&shutdown_threads) != NULL) {
+		struct shutdown_thread *thread;
+
+		/* Wake up all sleeping threads. */
+		SYS_SLIST_FOR_EACH_CONTAINER(&shutdown_threads, thread, node) {
+			k_sem_give(&thread->sem);
+		}
+
+		/* PM done after servicing interrupt for best latency */
+		ISR_DIRECT_PM();
```
[the remark this answers is not in the document]
Comment (evenl) [reply lib/bsdlib/bsdlib.c:65]: ok

File: lib/bsdlib/bsdlib.c [hunk 7]
```diff
@@ -39,9 +53,42 @@ static int _bsdlib_init(struct device *unused)
 		return 0;
 	}
 
+	if (sys_slist_peek_head(&shutdown_threads) != NULL) {
+		struct shutdown_thread *thread;
+
+		/* Wake up all sleeping threads. */
+		SYS_SLIST_FOR_EACH_CONTAINER(&shutdown_threads, thread, node) {
```
[the remark this answers is not in the document]
Comment (evenl) [reply lib/bsdlib/bsdlib.c:60]: yes and yes :)

File: lib/at_cmd/at_cmd.c (hunk 3 above)
Comment (evenl) [reply lib/at_cmd/at_cmd.c:150]: Because of the linebreak, it's easier to read with a empty line between.

File: lib/at_cmd/at_cmd.c (hunk 2 above)
Comment (evenl) [reply lib/at_cmd/at_cmd.c]: Removed this change

File: lib/bsdlib/nrf91_sockets.c [hunk 8]
```diff
@@ -406,6 +406,8 @@ static int nrf_to_z_dns_error_code(int nrf_error)
 		return DNS_EAI_NONAME;
 	case NRF_EINPROGRESS:
 		return DNS_EAI_INPROGRESS;
+	case NRF_ENETUNREACH:
+		return DNS_EAI_SERVICE;
```
Comment (rlubos) [inline lib/bsdlib/nrf91_sockets.c:410]: Is this a good error code here? `DNS_EAI_SERVICE` is directly related to the `service` argument of the `getaddrinfo` call (and indicates it was not a valid one). IMO `ENETUNREACH` falls under `DNS_EAI_SYSTEM` (we should set the `errno` variable in that case though, which from what I see is not done).

File: subsys/net/lib/download_client/src/download_client.c [hunk 9]
```diff
@@ -457,7 +461,14 @@ void download_thread(void *client, void *a, void *b)
 				/* Restart and suspend */
 				break;
 			}
-			reconnect(dl);
+
+			err = reconnect(dl);
+			if (err == -ENETUNREACH) {
+				LOG_WRN("Network is down, "
```
Comment (rlubos) [inline subsys/net/lib/download_client/src/download_client.c:467]: Shall we do this? I guess `ENETUNREACH` is just a temporary condition? Maybe we should just put it into sleep for some time and try again later? @lemrey

File: include/modem/bsdlib.h (hunk 4 above)
[the remark this answers is not in the document]
Comment (rlubos) [reply include/modem/bsdlib.h:25]: What about the docs?

File: lib/bsdlib/bsdlib.c (hunk 7 above)
[the remark this answers is not in the document]
Comment (rlubos) [reply lib/bsdlib/bsdlib.c:60]: Still not addressed

File: drivers/gps/nrf9160_gps/nrf9160_gps.c [hunk 10]
```diff
@@ -197,6 +199,22 @@ static void on_fix(struct device *dev)
 	}
 }
 
+static int open_socket(struct gps_drv_data *drv_data)
+{
+	drv_data->socket = nrf_socket(NRF_AF_LOCAL, NRF_SOCK_DGRAM,
+				  NRF_PROTO_GNSS);
```
[the remark this answers is not in the document]
Comment (evenl) [reply drivers/gps/nrf9160_gps/nrf9160_gps.c:205]: ok

File: drivers/gps/nrf9160_gps/nrf9160_gps.c [hunk 11]
```diff
@@ -197,6 +199,22 @@ static void on_fix(struct device *dev)
...
+
+	if (drv_data->socket >= 0) {
+		LOG_DBG("GPS socket created, fd: %d", drv_data->socket);
+	} else {
+		LOG_ERR("Could not initialize socket, error: %d)",
```
[the remark this answers is not in the document]
Comment (evenl) [reply drivers/gps/nrf9160_gps/nrf9160_gps.c:210]: heh, yes, fixing.

File: drivers/gps/nrf9160_gps/nrf9160_gps.c [hunk 12]
```diff
@@ -225,7 +243,27 @@ static void gps_thread(int dev_ptr)
...
 			}
 
-			LOG_ERR("recv() returned error: %d", len);
+			if (errno == EHOSTDOWN) {
+				LOG_DBG("GPS host is going down, sleeping");
+				k_delayed_work_cancel(&drv_data->timeout_work);
+				k_delayed_work_cancel(&drv_data->start_work);
+				atomic_clear(&drv_data->is_active);
+				atomic_set(&drv_data->is_shutdown, 1);
+
```
[the remark this answers is not in the document]
Comment (evenl) [reply drivers/gps/nrf9160_gps/nrf9160_gps.c:252]: Strange, I thought I did, fixing.

File: drivers/gps/nrf9160_gps/nrf9160_gps.c [hunk 13]
```diff
@@ -499,12 +546,25 @@ static int start(struct device *dev, struct gps_config *cfg)
...
 				NRF_SO_GNSS_FIX_RETRY,
 				&gps_cfg.retry,
 				sizeof(gps_cfg.retry));
-	if (retval != 0) {
+
+	if ((retval == -1) && ((errno == EFAULT) || (errno == EBADF))) {
+		LOG_WRN("Failed to set fix retry value, "
+			"will try to re-init GPS service");
+
+		if (open_socket(drv_data) != 0) {
```
[the remark this answers is not in the document]
Comment (evenl) [reply drivers/gps/nrf9160_gps/nrf9160_gps.c:561]: fixing.

File: lib/bsdlib/nrf91_sockets.c (hunk 8 above)
Comment (evenl) [reply lib/bsdlib/nrf91_sockets.c:410]: Yes, I agree, I misset the part in the getaddrinfo documentation that it can return errnos. Fixing

File: subsys/net/lib/download_client/src/download_client.c (hunk 9 above)
Comment (evenl) [reply subsys/net/lib/download_client/src/download_client.c:467]: In the case of ENETUNREACH the download client will spinnloop. I think it is better that the application takes care of this than trying to add some smart logic here. We don't know how long this intermediate condition is.

File: include/modem/bsdlib.h (hunk 4 above)
[the remark this answers is not in the document]
Comment (evenl) [reply include/modem/bsdlib.h:25]: Fixed.

File: lib/bsdlib/bsdlib.c (hunk 7 above)
[the remark this answers is not in the document]
Comment (evenl) [reply lib/bsdlib/bsdlib.c]: Fixed now, changed to SYS_LIST_FOR_EACH_CONTAINER_SAFE()

**Comment by evenl** [comment]
I think I have addressed all comments now, the PR should be ready.

File: drivers/gps/nrf9160_gps/nrf9160_gps.c [hunk 14]
```diff
@@ -225,7 +243,28 @@ static void gps_thread(int dev_ptr)
...
 			}
 
-			LOG_ERR("recv() returned error: %d", len);
+			if (errno == EHOSTDOWN) {
+				LOG_DBG("GPS host is going down, sleeping");
+				k_delayed_work_cancel(&drv_data->timeout_work);
+				k_delayed_work_cancel(&drv_data->start_work);
+				atomic_clear(&drv_data->is_active);
+				atomic_set(&drv_data->is_shutdown, 1);
+				close(drv_data->socket);
```
Comment (rlubos) [inline drivers/gps/nrf9160_gps/nrf9160_gps.c:252]: We use `nrf_socket` API here, so it should be `nrf_close()`. It makes difference.

File: drivers/gps/nrf9160_gps/nrf9160_gps.c [hunk 15]
```diff
@@ -499,12 +547,26 @@ static int start(struct device *dev, struct gps_config *cfg)
...
 				NRF_SO_GNSS_FIX_RETRY,
 				&gps_cfg.retry,
 				sizeof(gps_cfg.retry));
-	if (retval != 0) {
+
+	if ((retval == -1) && ((errno == EFAULT) || (errno == EBADF))) {
+		LOG_WRN("Failed to set fix retry value, "
+			"will try to re-init GPS service");
+
+		close(drv_data->socket);
```
Comment (rlubos) [inline drivers/gps/nrf9160_gps/nrf9160_gps.c:562]: Same here, `nrf_close()`

File: subsys/net/lib/download_client/src/download_client.c (hunk 9 above)
Comment (rlubos) [reply subsys/net/lib/download_client/src/download_client.c:467]: In that case we should notify the application with `error_evt_send(dl, ENETUNREACH);` so it's aware something fishy is going on. How will the application know when to resume?

File: subsys/net/lib/download_client/src/download_client.c [hunk 16]
```diff
@@ -540,12 +551,23 @@ void download_thread(void *client, void *a, void *b)
...
 			if (rc) {
 				/* Restart and suspend */
 				break;
 			}
-			reconnect(dl);
+			err = reconnect(dl);
+			if (err == ENETUNREACH) {
+				LOG_WRN("Network is down, "
+					"killing download thread");
+				return;
```
Comment (rlubos) [inline subsys/net/lib/download_client/src/download_client.c:569]: Here we should also suspend the thread instead of killing it. And notify an event.

File: drivers/gps/nrf9160_gps/nrf9160_gps.c (hunk 14 above)
Comment (evenl) [reply drivers/gps/nrf9160_gps/nrf9160_gps.c:252]: yes, that is true, there is no gnss socket in zephyr.

File: drivers/gps/nrf9160_gps/nrf9160_gps.c (hunk 15 above)
Comment (evenl) [reply drivers/gps/nrf9160_gps/nrf9160_gps.c:562]: fixed

File: subsys/net/lib/download_client/src/download_client.c (hunk 16 above)
Comment (evenl) [reply subsys/net/lib/download_client/src/download_client.c:569]: Though I had updated, that text, fixing.

File: drivers/gps/nrf9160_gps/nrf9160_gps.c [hunk 17]
```diff
@@ -225,7 +243,28 @@ static void gps_thread(int dev_ptr)
...
 			}
 
-			LOG_ERR("recv() returned error: %d", len);
+			if (errno == EHOSTDOWN) {
+				LOG_DBG("GPS host is going down, sleeping");
+				k_delayed_work_cancel(&drv_data->timeout_work);
+				k_delayed_work_cancel(&drv_data->start_work);
+				atomic_clear(&drv_data->is_active);
+				atomic_set(&drv_data->is_shutdown, 1);
+				nrf_socket(drv_data->socket);
```
Comment (rlubos) [inline drivers/gps/nrf9160_gps/nrf9160_gps.c:252]: You meant `nrf_close` right?

File: subsys/net/lib/download_client/src/download_client.c (hunk 9 above)
Comment (rlubos) [reply subsys/net/lib/download_client/src/download_client.c:465]: You forgot to add `error_evt_send(dl, ENETUNREACH);` here as well.

**Comment by evenl** [comment]
Really, omph!!! I am racing so much to get stuff done before the holidays, sorry about that, fixed.

File: subsys/net/lib/download_client/src/download_client.c [hunk 18]
```diff
@@ -427,6 +428,7 @@ void download_thread(void *client, void *a, void *b)
 			   sizeof(dl->buf) - dl->offset, 0);
 
 		if ((len == 0) || (len == -1)) {
+			int err;
```
[the remark this answers is not in the document]
Comment (evenl) [reply subsys/net/lib/download_client/src/download_client.c:431]: ok

File: subsys/net/lib/download_client/src/download_client.c [hunk 19]
```diff
@@ -540,12 +550,24 @@ void download_thread(void *client, void *a, void *b)
...
+				continue;
+			}
+
 			rc = error_evt_send(dl, ECONNRESET);
 			if (rc) {
 				/* Restart and suspend */
 				break;
 			}
-			reconnect(dl);
+			err = reconnect(dl);
```
[the remark this answers is not in the document]
Comment (evenl) [reply subsys/net/lib/download_client/src/download_client.c:564]: ok

File: subsys/net/lib/download_client/src/download_client.c [hunk 20]
```diff
@@ -686,9 +713,11 @@ void download_client_pause(struct download_client *client)
 	k_thread_suspend(client->tid);
 }
 
-void download_client_resume(struct download_client *client)
+int download_client_resume(struct download_client *client)
```
[the remark this answers is not in the document]
Comment (evenl) [reply subsys/net/lib/download_client/src/download_client.c:716]: Don't remember, reverting.
